### ISSUE TITLE
[DX] Form API: fields of type `help` and `markup` do not support `#attributes` or `#states`

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#3569]
I came across this while woking on #3075.

So this does not work:

```php
  $form['help'] = array(
    '#type' => 'help',
    '#markup' => t('Separate multiple classes with spaces.'),
    '#attributes' => array('class' => array('description')),
    '#states' => array(
      'invisible' => array(
        ':input[name="element"]' => array('value' => ''),
      ),
    ),
  );
```

...in that the class is not added, and the #states functionality never kicks in.

**Comment by klonos** [comment]
Seems that the same is true for `#markup` fields. So this does not work either:

```php
  $form['help'] = array(
    '#markup' => t('Separate multiple classes with spaces.'),
    '#attributes' => array('class' => array('description')),
    '#states' => array(
      'visible' => array(
        ':input[name="element"]' => array('value' => ''),
      ),
    ),
  );
```

**Comment by klonos** [comment]
...seems that a common way to solve this is to wrap `#markup` fields in a `'#type' => 'container'`, and then use `#states` on the container instead. This works, but there is extraneous markup, so not ideal.

ref:

https://knackforge.com/blog/sivaji/drupal-7-form-api-using-states-markup-field

http://drupal7steven.blogspot.com/2015/03/drupal-states-doesnt-work-on-markup.html

**Comment by klonos** [comment]
...'#type' => 'item' works (was "recently" fixed in D7, and we inherited the fix): https://www.drupal.org/project/drupal/issues/783438

Currently, the above-mentioned issue in d.org is marked as "needs work", because:

> Maybe it should be documented in the form api that #states does not work for elements of type 'markup'.
> Currently the form api says that #states can be used by 'any element':
> http://api.drupal.org/api/drupal/developer!topics!forms_api_reference.html/7#states

**Comment by klonos** [comment]
I have yet again come across these limitations while working on #1895...

So the goal was to:

1. Have a checkbox field that would act as a toggle to show/hide a text field
2. The label of the text field would be hidden (`'#title_display' => 'invisible'`), since the label of the checkbox/toggle would serve as a label.
3. The same help text should be displayed for both the checkbox and the textfield, but only once.

Original code, where there is only the text field - no checkbox/toggle:

```php
$form['text'] = array(
  '#type' => 'textarea',
  '#title' => t('Text area'),
  '#default_value' => $default_value,
  '#description' => t('This is help.'),
);
```

![Screen Shot 2019-05-26 at 9 06 52 pm](https://user-images.githubusercontent.com/2423362/58380867-3b99d680-7ffa-11e9-83fa-ae9a6763c39e.png)

...add the checkbox:

```php
$form['text_enable'] = array(
  '#type' => 'checkbox',
  '#title' => t('Provide text'),
);
$form['text'] = array(
  '#type' => 'textarea',
  '#title' => t('Text area'),
  '#default_value' => $default_value,
  '#description' => t('This is help.'),
);
```

![Screen Shot 2019-05-26 at 9 11 18 pm](https://user-images.githubusercontent.com/2423362/58380935-3db06500-7ffb-11e9-9ee8-53f5cdab9c78.png)

...hide the label of the text area:

```php
$form['text_enable'] = array(
  '#type' => 'checkbox',
  '#title' => t('Provide text'),
);
$form['text'] = array(
  '#type' => 'textarea',
  '#title' => t('Text area'),
  '#title_display' => 'invisible',
  '#default_value' => $default_value,
  '#description' => t('This is help.'),
);
```

![Screen Shot 2019-05-26 at 9 11 44 pm](https://user-images.githubusercontent.com/2423362/58380968-95e76700-7ffb-11e9-9580-8e72073f9126.png)

...use `#states` to hide the text field when the checkbox is not ticked:

```php
$form['text_enable'] = array(
  '#type' => 'checkbox',
  '#title' => t('Provide text'),
);
$form['text'] = array(
  '#type' => 'textarea',
  '#title' => t('Text area'),
  '#title_display' => 'invisible',
  '#default_value' => $default_value,
  '#description' => t('This is help.'),
  '#states' => array(
    'visible' => array(
      ':input[name="description_enable"]' => array('checked' => TRUE),
    ),
  ),
);
```

![Screen Shot 2019-05-26 at 9 12 38 pm](https://user-images.githubusercontent.com/2423362/58380971-a4358300-7ffb-11e9-821d-69e30c207945.png)

...the problem now is that the checkbox has no help text:

![Screen Shot 2019-05-26 at 9 12 30 pm](https://user-images.githubusercontent.com/2423362/58380970-a0a1fc00-7ffb-11e9-919b-4307db3499f6.png)

...add the help text to the checkbox:


```php
$form['text_enable'] = array(
  '#type' => 'checkbox',
  '#title' => t('Provide text'),
  '#description' => t('This is help.'),
);
$form['text'] = array(
  '#type' => 'textarea',
  '#title' => t('Text area'),
  '#title_display' => 'invisible',
  '#default_value' => $default_value,
  '#description' => t('This is help.'),
  '#states' => array(
    'visible' => array(
      ':input[name="description_enable"]' => array('checked' => TRUE),
    ),
  ),
);
```

![Screen Shot 2019-05-26 at 9 22 15 pm](https://user-images.githubusercontent.com/2423362/58381059-6ab14780-7ffc-11e9-81d6-7d49e5d1ed13.png)

...but when ticking the checkbox, you end up with duplicate help text:

![Screen Shot 2019-05-26 at 9 21 14 pm](https://user-images.githubusercontent.com/2423362/58381071-874d7f80-7ffc-11e9-9c24-72726d63cbb7.png)

Now, `#states` does not work on `#description`, so the following code is not valid, and you'll get errors:

```php
$form['text_enable'] = array(
  '#type' => 'checkbox',
  '#title' => t('Provide text'),
);
$form['text_enable']['#description'] = t('This is help.');
$form['text_enable']['#description']['#states'] = array(
  ...
);
```

The workaround to this, is to use a separate form element to hold the help text for the checkbox. `'#type' => 'help'` and plain `'#markup'` do not support `#states` (this issue here), so the second workaround is to use `'#type' => 'item'`:

```php
$form['text_enable'] = array(
  '#type' => 'checkbox',
  '#title' => t('Provide text'),
);
$form['text_enable_description'] = array(
  '#type' => 'item',
  '#markup' => t('This is help.'),
  '#states' => array(
    'visible' => array(
      ':input[name="description_enable"]' => array('checked' => FALSE),
    ),
  ),
);
$form['text'] = array(
  '#type' => 'textarea',
  '#title' => t('Text area'),
  '#title_display' => 'invisible',
  '#default_value' => $default_value,
  '#description' => t('This is help.'),
  '#states' => array(
    'visible' => array(
      ':input[name="description_enable"]' => array('checked' => TRUE),
    ),
  ),
);
```

This works (although that text is not styled as help text - we'll get to that):

![Screen Shot 2019-05-26 at 9 33 33 pm](https://user-images.githubusercontent.com/2423362/58381235-0abba080-7ffe-11e9-8679-51f4a7c56e00.png)

...and when the checkbox is ticked, the help text is hidden (so that there is no duplication with the existing help text in the text field):

![Screen Shot 2019-05-26 at 9 33 43 pm](https://user-images.githubusercontent.com/2423362/58381244-33439a80-7ffe-11e9-8f45-32e56b513a62.png)

OK, now the only thing that remains is to style that help text for the checkbox. Unfortunately, although `'#type' => 'item'` supports `#states`, it does not support `#attributes`, so this does not work:

```php
$form['text_enable'] = array(
  '#type' => 'checkbox',
  '#title' => t('Provide text'),
);
$form['text_enable_description'] = array(
  '#type' => 'item',
  '#markup' => t('This is help.'),
  '#attributes' => array('class' => array('description')),
  '#states' => array(
    'visible' => array(
      ':input[name="description_enable"]' => array('checked' => FALSE),
    ),
  ),
);
$form['text'] = array(
  '#type' => 'textarea',
  '#title' => t('Text area'),
  '#title_display' => 'invisible',
  '#default_value' => $default_value,
  '#description' => t('This is help.'),
  '#states' => array(
    'visible' => array(
      ':input[name="description_enable"]' => array('checked' => TRUE),
    ),
  ),
);
```

...so yet another workaround is to use `#prefix`/`#suffix` 🙄 :

```php
$form['text_enable'] = array(
  '#type' => 'checkbox',
  '#title' => t('Provide text'),
);
$form['text_enable_description'] = array(
  '#type' => 'item',
  '#prefix' => '<div class="description">',
  '#markup' => t('This is help.'),
  '#attributes' => array('class' => array('description')),
  '#suffix' => '</div>',
  '#states' => array(
    'visible' => array(
      ':input[name="description_enable"]' => array('checked' => FALSE),
    ),
  ),
);
$form['text'] = array(
  '#type' => 'textarea',
  '#title' => t('Text area'),
  '#title_display' => 'invisible',
  '#default_value' => $default_value,
  '#description' => t('This is help.'),
  '#states' => array(
    'visible' => array(
      ':input[name="description_enable"]' => array('checked' => TRUE),
    ),
  ),
);
```

This finally works, and the help text is styled as such, but the generated markup for the help text of the checkbox has an additional level of nesting:

```html
<div class="description">
  <div id="edit-text-enable-description" class="form-item form-type-item" style="display: block;">
   This is help.
  </div>
</div>
```

...so there is a gap between the field and the help text:

![Screen Shot 2019-05-26 at 9 46 45 pm](https://user-images.githubusercontent.com/2423362/58381390-e660c380-7fff-11e9-9cdf-f3be89a668ce.png)

Why does all this need to be so hard? 😓

**Comment by klonos** [comment]
This took **a lot** of testing/failure, and googling around for solutions. Ideally, I would expect this to work (`#states` works with '#type' => 'help'):

```php
$form['text_enable_description'] = array(
  '#type' => 'help',
  '#markup' => t('This is help.'),
  '#states' => array(
    'visible' => array(
      ':input[name="description_enable"]' => array('checked' => FALSE),
    ),
  ),
);
```

...or I would compromise with this at the very least (`#attributes` works with `'#type' => 'item'` - separate issue though):

```php
$form['text_enable_description'] = array(
  '#type' => 'item',
  '#attributes' => array('class' => array('description')),
  '#markup' => t('This is help.'),
  '#states' => array(
    'visible' => array(
      ':input[name="description_enable"]' => array('checked' => FALSE),
    ),
  ),
);
```

**Comment by BWPanda** [comment]
> So the question is, do we update documentation, or fix things so that #attributes and #states are supported for these types of fields?

I'd prefer making them work where/as expected, especially since the documentation indicates as much. Unless there's a technical reason why not...?

**Comment by oadaeh** [comment]
I wonder if this is related to #3489.
